accounts-db/src/accounts.rs

```diff
@@ -621,26 +669,42 @@ impl Accounts {
                 Err(err) => Err(err),
             })
             .collect();
-        self.lock_accounts_inner(tx_account_locks_results)
+        let (res, _) =
+            self.lock_accounts_inner(tx_account_locks_results, allow_self_conflicting_entries);
+        res
     }
 
     #[must_use]
     fn lock_accounts_inner(
         &self,
         tx_account_locks_results: Vec<Result<TransactionAccountLocks>>,
-    ) -> Vec<Result<()>> {
+        allow_self_conflicting_entries: &bool,
+    ) -> (Vec<Result<()>>, bool) {
         let account_locks = &mut self.account_locks.lock().unwrap();
-        tx_account_locks_results
-            .into_iter()
-            .map(|tx_account_locks_result| match tx_account_locks_result {
-                Ok(tx_account_locks) => self.lock_account(
-                    account_locks,
-                    tx_account_locks.writable,
-                    tx_account_locks.readonly,
-                ),
-                Err(err) => Err(err),
-            })
-            .collect()
+        let mut entry_accts_lookup = EntryAcctLocks::with_capacity(ENTRY_ACCTS_LOOKUP_TABLE_SIZE);
```

Comment:
why overallocate if we are re-allocating every batch?

either we should be re-using this allocation (thread-local maybe?) OR we should allocate enough space for this specific batch.

Reply:
@ryoqun might have some thoughts on this! he and I are pretty keen on avoiding any unneccessary allocations in this path

Reply:
One option could be that while we allocate memory for `tx_account_locks_results`, we estimate the capacity for `entry_accts_lookup`

we could do something like this in the `lock_accounts` function.

```suggestion
       let mut entry_lookup_capacity = 0;
        let tx_account_locks_results: Vec<Result<_>> = txs
            .map(|tx| { 
                let tx_account_locks = tx.get_account_locks(tx_account_lock_limit);
                if let Ok(tx_account_locks) = &tx_account_locks {
                    entry_lookup_capacity += tx_account_locks.readonly.len() + tx_account_locks.writable.len();
                }
                tx_account_locks
            })
            .collect();
        let mut entry_accts_lookup = EntryAcctLocks::with_capacity(entry_lookup_capacity);
        self.lock_accounts_inner(tx_account_locks_results, allow_self_conflicting_entries, &mut entry_accts_lookup)
```

Reply:
moreover, i don't think we should modify these critical code path aggressively.

Rather, maybe introduce ExecutionUnit or Subbatch, which isn't self-conflicting, as items of the `TransactionBatch`, which can be self-conflicting now.

Reply:
I think I disagree with this sentiment. We need to pay attention with these modifications, but adding sub-batches will lose us any efficiency gained by batching in the first place.

All of this needs to be pretty aggressively perf tested.

Reply:
what do guys think about the above suggestion to better estimate the size of `entry_accts_lookup` while we are constructing the `tx_account_locks_results`?

Reply:
From my perspective the most important thing here (after correctness) is performance, so I'd lean towards something along lines of:
```rust
thread_local!{
	static ENTRY_ACCT_LOCKS: EntryAcctLocks = EntryAcctLocks::with_capacity(entry_lookup_capacity); 
}
```

this means every thread that is calling this code will have it's own struct to look up in. Just clear when we use it in this fn. On most calls we will never have to allocate; if we exceed the cap we hit allocation only that time.
Down-side of course being this memeory will live as long as the threads that call into it.

@ryoqun with your new code-path would that mean it's each handler thread or each is this done by the scheduler thread?

w/ 16k elements this is about 500K bytes per thread I think